### PR TITLE
CLI infrastructure for semantic analysis compiler stage

### PR DESCRIPTION
**Pull request by problame**

### CODE REVIEW

File: src/sem.rs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+use failure::{format_err, Error, Fail};
+use std::{collections::HashMap, rc::Rc};
+
+type MethodDefinitions<'a, 'f> = HashMap<
+    (Symbol, Symbol, ast::ClassMemberKindDiscriminants),
+    &'a Spanned<'f, ast::ClassMember<'f>>,
+>;
+
+pub fn check<'a, 'f>(_ast: &'a ast::AST<'f>, context: &context::Context<'_>) -> Result<(), Error> {
+    let mut member_definitions: MethodDefinitions<'a, 'f> = HashMap::new();
```
Comment (flip1995) [inline src/sem.rs]: I would write visitors which need to store some information as:

```rust
struct WhatAmIDoingVisitor {
    member_defs: MethodDefinitions,
}

impl WhatAmIDoingVisitor {
    first_pass(&mut self, n: NodeKind, cx: Context) {
        ...
    }
}
```

Comment (problame) [reply src/sem.rs]: I will make a separate PR with the POC visitor. This branch got out of hand, will force push something that fits the title of this PR.

**Pull request merged**